language-plutus-core/stdlib/Language/PlutusCore/StdLib/Meta/Data/Tuple.hs

```diff
@@ -3,104 +3,163 @@
+--         pure $ VarDecl () n $ mkTyVar () $ tyVars !! i
+
+--     caseArg <- liftQuote $ freshName () "case"
+--     let caseTy = mkIterTyFun () (fmap (mkTyVar ()) tyVars) (TyVar () resultType)
+--     pure $
+--         -- /\T_1 .. T_n
+--         mkIterTyAbs () tyVars $
+--         -- \arg_1 .. arg_n
+--         mkIterLamAbs () args $
+--         -- /\R
+--         TyAbs () resultType (Type ()) $
+--         -- \case
+--         LamAbs () caseArg caseTy $
+--         -- case arg_1 .. arg_n
+--         mkIterApp () (Var () caseArg) $ fmap (mkVar ()) args
+
+-- -- | Given an arity @n@ and an index @i@, create a function for accessing the i'th component of a n-tuple.
+-- --
+-- -- @
+-- --     /\(T_1 :: *) .. (T_n :: *) .
+-- --         \(tuple : all (R :: *) . (T_1 -> .. -> T_n -> R) -> R)) .
+-- --             tuple {T_i} (\(arg_1 : T_1) .. (arg_n : T_n) . arg_n)
+-- -- @
+-- getBuiltinTupleAccessor :: MonadQuote m => Int -> Int -> m (Term TyName Name ())
+-- getBuiltinTupleAccessor arity index = rename =<< do
+--     tyVars <- for [0..(arity-1)] $ \i -> do
+--         tn <- liftQuote $ freshTyName () $ strToBs $ "t_" ++ show i
+--         pure $ TyVarDecl () tn $ Type ()
+
+--     tupleTy <- do
+--         genericTuple <- getBuiltinTuple arity
+--         pure $ mkIterTyApp () genericTuple (fmap (mkTyVar ()) tyVars)
+--     let selectedTy = mkTyVar () $ tyVars !! index
+
+--     args <- for [0..(arity -1)] $ \i -> do
+--         n <- liftQuote $ freshName () $ strToBs $ "arg_" ++ show i
+--         pure $ VarDecl () n $ mkTyVar () $ tyVars !! i
+--     let selectedArg = mkVar () $ args !! index
+
+--     tupleArg <- liftQuote $ freshName () "tuple"
+--     pure $
+--         -- /\T_1 .. T_n
+--         mkIterTyAbs () tyVars $
+--         -- \tuple :: (tupleN T_1 .. T_n)
+--         LamAbs () tupleArg tupleTy $
+--         -- tuple {T_i}
+--         Apply () (TyInst () (Var () tupleArg) selectedTy) $
+--         -- \arg_1 .. arg_n . arg_i
+--         mkIterLamAbs () args selectedArg
+
+data Tuple ann = Tuple
+    { _tupleElementTypes :: [Type TyName ann]
+    , _tupleTerm         :: Term TyName Name ann
+    }
+
+getTupleType :: MonadQuote m => ann -> Tuple ann -> m (Type TyName ann)
+getTupleType ann (Tuple elTys _) = liftQuote $ do
+    resultTy <- freshTyName ann "r"
+    let caseTy = mkIterTyFun ann elTys (TyVar ann resultTy)
+    pure
+        . TyForall ann resultTy (Type ann)
+        . TyFun ann caseTy
+        $ TyVar ann resultTy
+
+tupleTypeTermAt
+    :: MonadQuote m => ann -> Int -> Tuple ann -> m (Type TyName ann, Term TyName Name ann)
+tupleTypeTermAt ann ind (Tuple elTys term) = liftQuote $ do
+    args <- ifor elTys $ \i ty -> do
+        n <- freshName ann $ strToBs $ "arg_" ++ show i
+        pure $ VarDecl ann n ty
+    let selectedTy  = elTys !! ind
+        selectedArg = mkVar ann $ args !! ind
+
+    pure
+        ( selectedTy
+        , Apply ann (TyInst ann term selectedTy) $ mkIterLamAbs ann args selectedArg
```

Comment:
This is going to duplicate the entire tuple every time you access an element, no?

Reply:
Yes, just like an application of `getBuiltinTupleAccessor` did previously. This is not a problem as long as `term` is a variable. I.e. this is no different from the previous solution, but needs to be documented properly.